### PR TITLE
🌱 add validation for default, enum and example definitions in variables

### LINKED ISSUE #5831: ClusterClass: schema validation of Default, Enum, Example
Currently, we only validate that `Default`, `Enum` and `Example` (in `JSONSchemaProps`) are valid JSON. We don't validate that they comply to the schema.

Goal of this issue is to discuss if we also want to validate those fields.

In my opinion, the most important field of those is `Default`. If `Default` does not comply to the schema and a variable is defaulted, the resulting Cluster will automatically be invalid. This is an especially bad UX as the user who created the Cluster didn't set the variable and potentially didn't write the ClusterClass, but the Cluster creation/update still fails because of a value we set during defaulting in the webhook.

Notes:
* If we do it, we have to do this recursively also for all nested fields (Objects/Arrays)
* Not sure what the performance impact is, but it's probably negligible as we would only run this validation during ClusterClass create and update

Notes on implementation:
* I think it could be as easy as taking the validator we create [here](https://github.com/kubernetes-sigs/cluster-api/blob/2212968e5b962c66d39f051791bce9e3e8a0e2f9/internal/topology/variables/clusterclass_variable_validation.go#L191) and using it directly below to validate `Default`, `Example` and all `Enum` values individually + adding unit tests
    * Note: This function is already called recursively.

Context:
Afaik Kubernetes is validating neither of those fields in CRDs (but I don't think that should influence us too much)

/kind feature
/area topology

### PR DESCRIPTION
**Pull request by killianmuldoon**
This PR adds schema validation for values defined for default, enum and example in a ClusterClass variable. 

e.g. 

If a user defines a schema with an integer and minimum value 1 a default, example or enum member of -100 will be found to be invalid and the ClusterClass will be rejected on creation or update.

It also removes the Kubernetes CRD default validation that was part of the code previously.

Fixes #5831

### CODE REVIEW

File: internal/topology/variables/clusterclass_variable_validation.go [hunk 1]
```diff
@@ -178,20 +152,43 @@ func validateSchema(schema *clusterv1.JSONSchemaProps, fldPath *field.Path) fiel
 		return append(allErrs, validationErrors...)
 	}
 
-	// Validate defaults in the structural schema.
-	validationErrors, err := structuraldefaulting.ValidateDefaults(fldPath.Child("schema"), ss, true, true)
```
Comment (killianmuldoon) [inline internal/topology/variables/clusterclass_variable_validation.go:182]: This is a recursive function itself so each default value was being recursed twice producing one defaulting error (if there was one) for each level of nesting in the object)

Comment (sbueringer) [reply internal/topology/variables/clusterclass_variable_validation.go:182]: I think we should keep it but it shouldn't be part of our recursion

**Comment by sbueringer** [comment]
I think we should restructure. The current recursion is not ideal (before this PR)

~ something like:

* top-level validateRootSchema func:
    * convertToAPIExtensionsJSONSchemaProps
        * move the basic JSON validation into this func
    * do all the stuff with the Kubernetes schema validators including ValidateDefaulting
    * call ValidateSchema on the current top-level apiExtensionsSchema

* validateSchema:
        * validate type
        * build a validator with validation.NewSchemaValidator
        * validate Example & Enums
        * call validateSchema recursively for array/object

File: internal/topology/variables/cluster_variable_defaulting.go [hunk 2]
```diff
@@ -91,10 +91,9 @@ func defaultClusterVariable(clusterVariable *clusterv1.ClusterVariable, clusterC
...
 
 	// Convert schema to Kubernetes APIExtensions schema.
-	apiExtensionsSchema, err := convertToAPIExtensionsJSONSchemaProps(&clusterClassVariable.Schema.OpenAPIV3Schema)
-	if err != nil {
-		return nil, field.ErrorList{field.Invalid(fldPath, "",
-			fmt.Sprintf("invalid schema in ClusterClass for variable %q: error to convert schema %v", clusterVariable.Name, err))}
+	apiExtensionsSchema, errs := convertToAPIExtensionsJSONSchemaProps(&clusterClassVariable.Schema.OpenAPIV3Schema, fldPath)
+	if len(errs) > 0 {
+		return nil, errs
 	}
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_defaulting.go:97]:
```suggestion
	apiExtensionsSchema, errs := convertToAPIExtensionsJSONSchemaProps(&clusterClassVariable.Schema.OpenAPIV3Schema, field.NewPath("schema"))
	if len(errs) > 0 {
		return nil, field.ErrorList{field.Invalid(fldPath, "",
			fmt.Sprintf("invalid schema in ClusterClass for variable %q: error to convert schema %v", clusterClassVariable.Name, errs))}
	}
```

The fldPath here is the path of the variable. I think passing in a new pseudo path is the best we can do. This results in the following error message:
```yaml
[spec.topology.variables[0]: Invalid value: "": invalid schema in ClusterClass for variable "cpu": error to convert schema [schema.default: Invalid value: "1\"": default is not valid JSON: invalid character '"' after top-level value]]
```

Note: we have to use `clusterClassVariable.Name` not `clusterVariable.Name` my previous code would have paniced.

File: internal/topology/variables/cluster_variable_validation.go [hunk 3]
```diff
@@ -104,10 +104,9 @@ func ValidateClusterVariable(clusterVariable *clusterv1.ClusterVariable, cluster
...
 
 	// Convert schema to Kubernetes APIExtensions Schema.
-	apiExtensionsSchema, err := convertToAPIExtensionsJSONSchemaProps(&clusterClassVariable.Schema.OpenAPIV3Schema)
-	if err != nil {
-		return field.ErrorList{field.InternalError(fldPath,
-			fmt.Errorf("failed to convert schema definition for variable %q; ClusterClass should be checked: %v", clusterVariable.Name, err))} // TODO: consider if to add ClusterClass name
+	apiExtensionsSchema, allErrs := convertToAPIExtensionsJSONSchemaProps(&clusterClassVariable.Schema.OpenAPIV3Schema, fldPath)
+	if len(allErrs) > 0 {
+		return allErrs
 	}
```
Comment (sbueringer) [inline internal/topology/variables/cluster_variable_validation.go:110]:
```suggestion
	apiExtensionsSchema, allErrs := convertToAPIExtensionsJSONSchemaProps(&clusterClassVariable.Schema.OpenAPIV3Schema, field.NewPath("schema"))
	if len(allErrs) > 0 {
		return field.ErrorList{field.InternalError(fldPath,
			fmt.Errorf("failed to convert schema definition for variable %q; ClusterClass should be checked: %v", clusterClassVariable.Name, allErrs))} // TODO: consider if to add ClusterClass name
	}
```

Similar issue & solution

File: internal/topology/variables/schema.go [hunk 4]
```diff
@@ -95,5 +115,31 @@ func convertToAPIExtensionsJSONSchemaProps(schema *clusterv1.JSONSchemaProps) (*
...
 
-	return props, nil
+	if len(schema.Properties) > 0 {
+		props.Properties = map[string]apiextensions.JSONSchemaProps{}
+		for propertyName, propertySchema := range schema.Properties {
+			p := propertySchema
+			apiExtensionsSchema, err := convertToAPIExtensionsJSONSchemaProps(&p, fldPath.Child("properties").Key(propertyName))
+			if err != nil {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("properties").Key(propertyName), "",
+					fmt.Sprintf("failed to convert schema of property: %v", err)))
```
Comment (sbueringer) [inline internal/topology/variables/schema.go:125]:
```suggestion
					fmt.Sprintf("failed to convert schema: %v", err)))
```
I think that's better as the path is there and "of property" sounds a bit unclear

File: internal/topology/variables/schema.go [hunk 5]
```diff
@@ -95,5 +115,31 @@ func convertToAPIExtensionsJSONSchemaProps(schema *clusterv1.JSONSchemaProps) (*
...
+				props.Properties[propertyName] = *apiExtensionsSchema
+			}
+		}
+	}
+
+	if schema.Items != nil {
+		apiExtensionsSchema, err := convertToAPIExtensionsJSONSchemaProps(schema.Items, fldPath.Child("items"))
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("items"), "",
+				fmt.Sprintf("failed to convert schema of items: %v", err)))
```
Comment (sbueringer) [inline internal/topology/variables/schema.go:136]:
```suggestion
				fmt.Sprintf("failed to convert schema: %v", err)))
```
similar to above

File: internal/topology/variables/clusterclass_variable_validation.go [hunk 6]
```diff
@@ -193,6 +149,48 @@ func validateSchema(schema *clusterv1.JSONSchemaProps, fldPath *field.Path) fiel
 	}); err != nil {
 		return append(allErrs, field.Invalid(fldPath, "", fmt.Sprintf("failed to build validator: %v", err)))
 	}
+	allErrs = append(allErrs, validateSchema(apiExtensionsSchema, fldPath)...)
```
Comment (sbueringer) [inline internal/topology/variables/clusterclass_variable_validation.go:152]:
```suggestion

	allErrs = append(allErrs, validateSchema(apiExtensionsSchema, fldPath)...)
```
nit, just to separate it a bit from the schema validator above

File: internal/topology/variables/clusterclass_variable_validation.go [hunk 7]
```diff
@@ -193,6 +149,48 @@ func validateSchema(schema *clusterv1.JSONSchemaProps, fldPath *field.Path) fiel
...
+	return allErrs
+}
+
+func validateSchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// Validate that type is one of the validVariableTypes.
+	switch {
+	case len(schema.Type) == 0:
+		return field.ErrorList{field.Required(fldPath.Child("type"), "openAPIV3Schema.type cannot be empty")}
```
Comment (sbueringer) [inline internal/topology/variables/clusterclass_variable_validation.go:161]:
```suggestion
		return field.ErrorList{field.Required(fldPath.Child("type"), "type cannot be empty")}
```
Let's drop it here too, given that it's wrong on deeper layers

File: internal/topology/variables/clusterclass_variable_validation.go [hunk 8]
```diff
@@ -193,6 +149,48 @@ func validateSchema(schema *clusterv1.JSONSchemaProps, fldPath *field.Path) fiel
...
+	case !validVariableTypes.Has(schema.Type):
+		return field.ErrorList{field.NotSupported(fldPath.Child("type"), schema.Type, validVariableTypes.List())}
+	}
+
+	// validate enums and example here.
```
Comment (sbueringer) [inline internal/topology/variables/clusterclass_variable_validation.go:166]:
```suggestion
```

File: internal/topology/variables/clusterclass_variable_validation_test.go [hunk 9]
```diff
@@ -423,6 +421,240 @@ func Test_ValidateClusterClassVariable(t *testing.T) {
...
+			},
+			wantErr: true,
+		},
+		{
+			name: "pass if variable is an object with default value valid by the given schema",
+			clusterClassVariable: &clusterv1.ClusterClassVariable{
+				Name: "var",
+				Schema: clusterv1.VariableSchema{
+					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+						Type: "object",
```
Comment (sbueringer) [inline internal/topology/variables/clusterclass_variable_validation_test.go:457]: Cam we also add a default value on the top-level pls? (would be nice to also have a case where the top-level default value (which is a object) does not match the schema

Comment (sbueringer) [reply internal/topology/variables/clusterclass_variable_validation_test.go:457]: Similar for example and enum (enum can be objects too). Thank you!!!

File: internal/topology/variables/cluster_variable_defaulting.go (hunk 2 above)
Comment (killianmuldoon) [reply internal/topology/variables/cluster_variable_defaulting.go:97]: It seems strange to be wrapping one error list in another like this. Is the variableName the context you want to preserve here?

Comment (sbueringer) [reply internal/topology/variables/cluster_variable_defaulting.go:97]: My main motivation is getting a readable error :)

Without this change it looks like the schema is part of the Cluster. So imho nesting error lists is pretty similar to the usual error nesting and leads to similarly good results.

I would try to get an error which:
* a) express which variable is broken (aka contains the variable path)
* b) tells me which part of the schema is broken (aka tells me something about the location of the error in the schema)

Comment (sbueringer) [reply internal/topology/variables/cluster_variable_defaulting.go:97]: The current error looks like this:
```yaml
[spec.topology.variables[0].default: Invalid value: "1\"": default is not valid JSON: invalid character '"' after top-level value]
```
(it points to a non-existing field default on the variable)

File: internal/topology/variables/clusterclass_variable_validation.go [hunk 10]
```diff
@@ -194,5 +151,51 @@ func validateSchema(schema *clusterv1.JSONSchemaProps, fldPath *field.Path) fiel
...
+	validator, _, err := validation.NewSchemaValidator(&apiextensions.CustomResourceValidation{
+		OpenAPIV3Schema: schema,
+	})
+	if err != nil {
+		return append(allErrs, field.Invalid(fldPath, "", fmt.Sprintf("failed to build validator: %v", err)))
+	}
+
+	if schema.Example != nil {
+		if errs := validation.ValidateCustomResource(fldPath, *schema.Example, validator); len(errs) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("example"), schema.Example, fmt.Sprintf("invalid value in example %v", errs)))
```
Comment (sbueringer) [inline internal/topology/variables/clusterclass_variable_validation.go:179]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath.Child("example"), schema.Example, fmt.Sprintf("invalid value in example: %v", errs)))
```
nit

File: internal/topology/variables/schema.go [hunk 11]
```diff
@@ -43,48 +48,62 @@ func convertToAPIExtensionsJSONSchemaProps(schema *clusterv1.JSONSchemaProps) (*
...
+					allErrs = append(allErrs, field.Invalid(fldPath.Child("enum").Index(i), string(enum.Raw),
+						fmt.Sprintf("failed to convert enum value: %v", err)))
+				} else {
+					props.Enum = append(props.Enum, v)
+				}
+			}
 		}
-		props.Default = &v
 	}
+	var value apiextensions.JSON
```
Comment (sbueringer) [inline internal/topology/variables/schema.go:90]: Please move the value as v into the if again

File: internal/topology/variables/schema_test.go [hunk 12]
```diff
@@ -64,6 +65,10 @@ func Test_convertToAPIExtensionsJSONSchemaProps(t *testing.T) {
 		{
 			name: "pass for schema validation with enum & default",
 			schema: &clusterv1.JSONSchemaProps{
+
+				Example: &apiextensionsv1.JSON{
```
Comment (sbueringer) [inline internal/topology/variables/schema_test.go:69]:
```suggestion
				Example: &apiextensionsv1.JSON{
```
nit

File: internal/topology/variables/clusterclass_variable_validation_test.go [hunk 13]
```diff
@@ -423,6 +421,404 @@ func Test_ValidateClusterClassVariable(t *testing.T) {
...
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+
+		{
+			name: "pass on variable with an example that is invalid by the given schema",
```
Comment (sbueringer) [inline internal/topology/variables/clusterclass_variable_validation_test.go:551]:
```suggestion
			name: "pass on variable with an example that is valid by the given schema",
```

File: internal/topology/variables/clusterclass_variable_validation_test.go [hunk 14]
```diff
@@ -423,6 +421,404 @@ func Test_ValidateClusterClassVariable(t *testing.T) {
...
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "fail if variable is an object with an example invalidated by the given schema",
```
Comment (sbueringer) [inline internal/topology/variables/clusterclass_variable_validation_test.go:601]:
```suggestion
			name: "fail if variable is an object with an example valid by the given schema",
```

Comment (killianmuldoon) [reply internal/topology/variables/clusterclass_variable_validation_test.go:601]: This is correct - the object is invalid here.

Comment (sbueringer) [reply internal/topology/variables/clusterclass_variable_validation_test.go:601]: Ah mom

**Review by fabriziopandini — COMMENTED** [review]
changes lgtm, I will make another past on tests

File: internal/topology/variables/clusterclass_variable_validation_test.go (hunk 14 above)
Comment (sbueringer) [reply internal/topology/variables/clusterclass_variable_validation_test.go:601]: Wrong line :/ Should be l.793

Comment (killianmuldoon) [reply internal/topology/variables/clusterclass_variable_validation_test.go:601]: :) Fixed now.

**Comment by sbueringer** [comment]
lgtm pending squash

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**